core/src/services/onedrive/backend.rs

```diff
@@ -0,0 +1,212 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use async_trait::async_trait;
+use http::{header, Request, Response, StatusCode};
+use std::fmt::Debug;
+
+use crate::{
+    ops::{OpRead, OpWrite},
+    raw::{
+        build_rooted_abs_path, new_request_build_error, parse_into_metadata, percent_encode_path,
+        Accessor, AccessorInfo, AsyncBody, HttpClient, IncomingAsyncBody, RpRead, RpWrite,
+    },
+    types::Result,
+    Capability, Error, ErrorKind,
+};
+
+use super::{error::parse_error, writer::OneDriveWriter};
+
+#[derive(Clone)]
+pub struct OnedriveBackend {
+    root: String,
+    access_token: String,
+    client: HttpClient,
+}
+
+impl OnedriveBackend {
+    pub(crate) fn new(root: String, access_token: String, http_client: HttpClient) -> Self {
+        Self {
+            root,
+            access_token,
+            client: http_client,
+        }
+    }
+}
+
+impl Debug for OnedriveBackend {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let mut de = f.debug_struct("OneDriveBackend");
+        de.field("root", &self.root);
+        de.field("access_token", &self.access_token);
+        de.finish()
+    }
+}
+
+#[async_trait]
+impl Accessor for OnedriveBackend {
+    type Reader = IncomingAsyncBody;
+    type BlockingReader = ();
+    type Writer = OneDriveWriter;
+    type BlockingWriter = ();
+    type Pager = ();
+    type BlockingPager = ();
+
+    fn info(&self) -> AccessorInfo {
+        let mut ma = AccessorInfo::default();
+        ma.set_scheme(crate::Scheme::Onedrive)
+            .set_root(&self.root)
+            .set_capability(Capability {
+                read: true,
+                read_can_next: true,
+                write: true,
+                list: true,
+                copy: true,
+                rename: true,
+                ..Default::default()
+            });
+
+        ma
+    }
+
+    async fn read(&self, path: &str, _args: OpRead) -> Result<(RpRead, Self::Reader)> {
+        let resp = self.onedrive_get(path).await?;
+
+        let status = resp.status();
+
+        if status.is_redirection() {
```

Comment:
There're two requests for a single read Op because the first request will get http redirect.